File: moling-server/app/dao/secret_dao.py
"""墨灵 (Moling) — Secret DAO."""

from __future__ import annotations

from typing import Optional

from sqlalchemy import func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.secret import Secret


class SecretDAO(BaseDAO[Secret]):
    """Data access for Secret (秘密矩阵)."""

    def __init__(self) -> None:
        super().__init__(Secret)

    # ---- Read ----

    async def get_by_id(
        self,
        db: AsyncSession,
        secret_id: int,
    ) -> Optional[Secret]:
        """Get a single secret by primary key."""
        return await self.get(db, secret_id)

    async def list_by_project(
        self,
        db: AsyncSession,
        project_id: int,
    ) -> list[Secret]:
        """List all secrets for a project."""
        stmt = (
            select(Secret)
            .where(Secret.project_id == project_id)
            .order_by(Secret.id.asc())
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def list_by_secrecy_level(
        self,
        db: AsyncSession,
        project_id: int,
        level: str,
    ) -> list[Secret]:
        """List secrets filtered by secrecy_level (hidden / partial / revealed)."""
        stmt = (
            select(Secret)
            .where(
                Secret.project_id == project_id,
                Secret.secrecy_level == level,
            )
            .order_by(Secret.id.asc())
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    # ---- Write ----

    async def update_by_id(
        self,
        db: AsyncSession,
        secret_id: int,
        obj_in: dict,
    ) -> Optional[Secret]:
        """Update a secret by id, returning the refreshed object."""
        secret = await self.get(db, secret_id)
        if secret is None:
            return None
        for field, value in obj_in.items():
            if hasattr(secret, field) and field != "id":
                setattr(secret, field, value)
        await db.flush()
        await db.refresh(secret)
        return secret

    async def batch_create(
        self,
        db: AsyncSession,
        objs_in: list[dict],
    ) -> list[Secret]:
        """Batch create secrets, returning the created objects."""
        secrets = [Secret(**data) for data in objs_in]
        db.add_all(secrets)
        await db.flush()
        for s in secrets:
            await db.refresh(s)
        return secrets

    async def batch_update(
        self,
        db: AsyncSession,
        updates: list[dict],
    ) -> int:
        """Batch update secrets. Each dict must contain 'id' plus fields to update.
        Returns total number of rows updated.
        """
        total = 0
        for upd in updates:
            secret_id = upd.pop("id")
            stmt = (
                update(Secret)
                .where(Secret.id == secret_id)
                .values(**upd)
            )
            result = await db.execute(stmt)
            total += result.rowcount
        await db.flush()
        return total

    # ---- Aggregation ----

    async def calculate_debt_by_project(
        self,
        db: AsyncSession,
        project_id: int,
    ) -> int:
        """Sum the debt of all secrets in a project."""
        stmt = (
            select(func.coalesce(func.sum(Secret.debt), 0))
            .where(Secret.project_id == project_id)
        )
        result = await db.execute(stmt)
        return result.scalar_one()

    async def count_by_project(
        self,
        db: AsyncSession,
        project_id: int,
    ) -> int:
        """Count secrets in a project."""
        stmt = (
            select(func.count())
            .select_from(Secret)
            .where(Secret.project_id == project_id)
        )
        result = await db.execute(stmt)
        return result.scalar_one()


# Singleton
secret_dao = SecretDAO()

File: moling-server-rs/crates/moling-db/src/dao/secret_dao.rs
//! Secret DAO — secret (秘密矩阵) CRUD and batch operations.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, QuerySelect, Set};

use crate::entities::secret::{self, ActiveModel, Entity as Secret, Model as SecretModel};

/// Secret data access object.
#[derive(Clone, Default)]
pub struct SecretDao;

impl SecretDao {
    /// Find a single secret by primary key.
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: &str,
    ) -> AppResult<Option<SecretModel>> {
        Secret::find_by_id(id.to_owned())
            .filter(secret::Column::IsDeleted.eq(false))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Secret: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List all secrets for a project, ordered by id ascending.
    pub async fn list_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<Vec<SecretModel>> {
        Secret::find()
            .filter(secret::Column::ProjectId.eq(project_id))
            .filter(secret::Column::IsDeleted.eq(false))
            .order_by_asc(secret::Column::Id)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, "Secret: database error listing by project: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List secrets filtered by secrecy_level (hidden / partial / revealed).
    pub async fn list_by_secrecy_level(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        level: &str,
    ) -> AppResult<Vec<SecretModel>> {
        Secret::find()
            .filter(secret::Column::ProjectId.eq(project_id))
            .filter(secret::Column::SecrecyLevel.eq(level))
            .filter(secret::Column::IsDeleted.eq(false))
            .order_by_asc(secret::Column::Id)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, %level, "Secret: database error listing by level: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Create a new secret.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<SecretModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("Secret: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Update an existing secret.
    pub async fn update(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<SecretModel> {
        model.update(db).await.map_err(|e| {
            tracing::error!("Secret: database error updating: {e}");
            AppError::internal("Database update failed")
        })
    }

    /// Soft-delete a secret.
    pub async fn soft_delete(&self, db: &DatabaseConnection, id: &str) -> AppResult<()> {
        use sea_orm::IntoActiveModel;
        use chrono::Utc;

        let entity = self.find_by_id(db, id).await?.ok_or_else(|| {
            AppError::not_found("秘密不存在")
        })?;

        let mut active = entity.into_active_model();
        active.is_deleted = Set(true);
        active.deleted_at = Set(Some(Utc::now()));
        active.update(db).await.map_err(|e| {
            tracing::error!(%id, "Secret: database error soft-deleting: {e}");
            AppError::internal("Database update failed")
        })?;
        Ok(())
    }

    /// Batch create secrets, returning the created models.
    pub async fn batch_create(
        &self,
        db: &DatabaseConnection,
        models: Vec<ActiveModel>,
    ) -> AppResult<Vec<SecretModel>> {
        let mut results = Vec::with_capacity(models.len());
        for model in models {
            let created = model.insert(db).await.map_err(|e| {
                tracing::error!("Secret: database error batch creating: {e}");
                AppError::internal("Database insert failed")
            })?;
            results.push(created);
        }
        Ok(results)
    }

    /// Batch update secrets. Each model must have its primary key set.
    /// Returns the total number of rows updated.
    pub async fn batch_update(
        &self,
        db: &DatabaseConnection,
        updates: Vec<ActiveModel>,
    ) -> AppResult<u64> {
        let mut total: u64 = 0;
        for model in updates {
            model.update(db).await.map_err(|e| {
                tracing::error!("Secret: database error batch updating: {e}");
                AppError::internal("Database update failed")
            })?;
            total += 1;
        }
        Ok(total)
    }

    /// Sum the debt of all non-deleted secrets in a project.
    pub async fn calculate_debt_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<i32> {
        use sea_orm::entity::*;
        Secret::find()
            .filter(secret::Column::ProjectId.eq(project_id))
            .filter(secret::Column::IsDeleted.eq(false))
            .select_only()
            .column_as(secret::Column::Debt.sum(), "total_debt")
            .into_tuple::<(Option<i32>,)>()
            .one(db)
            .await
            .map(|r| r.and_then(|(v,)| v).unwrap_or(0))
            .map_err(|e| {
                tracing::error!(project_id, "Secret: database error calculating debt: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Count non-deleted secrets in a project.
    pub async fn count_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<u64> {
        Secret::find()
            .filter(secret::Column::ProjectId.eq(project_id))
            .filter(secret::Column::IsDeleted.eq(false))
            .count(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, "Secret: database error counting: {e}");
                AppError::internal("Database query failed")
            })
    }
}
